src/main/java/ee/carlrobert/codegpt/codecompletions/InlineCodeCompletionRenderer.java

```diff
@@ -0,0 +1,226 @@
+import com.intellij.openapi.util.Key;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.ui.JBColor;
+import ee.carlrobert.codegpt.completions.CallParameters;
+import ee.carlrobert.codegpt.completions.CompletionRequestHandler;
+import ee.carlrobert.codegpt.completions.ConversationType;
+import ee.carlrobert.codegpt.conversations.Conversation;
+import ee.carlrobert.codegpt.conversations.ConversationService;
+import ee.carlrobert.codegpt.conversations.message.Message;
+import ee.carlrobert.codegpt.settings.configuration.ConfigurationState;
+import ee.carlrobert.codegpt.settings.service.FillInTheMiddle;
+import ee.carlrobert.codegpt.settings.state.SettingsState;
+import java.awt.Font;
+import java.awt.FontMetrics;
+import java.awt.Graphics2D;
+import java.awt.event.KeyEvent;
+import java.awt.event.KeyListener;
+import java.awt.geom.Rectangle2D;
+import javax.swing.Timer;
+import org.jetbrains.annotations.NotNull;
+
+public class InlineCodeCompletionRenderer implements EditorCustomElementRenderer, DocumentListener,
+    KeyListener, CaretListener, EditorMouseListener, SelectionListener {
+
+  public static final Key<Inlay<InlineCodeCompletionRenderer>> INLAY_KEY = Key.create(
+      "codegpt.inlay");
+  public static final String ACTION_ID = "InsertInlineTextAction";
+  public static final int MAX_OFFSET = 400; // TODO: Add in settings? Always entire file?
+
+  private Timer typingTimer;
+  private String inlayText = "";
+  private CompletionRequestHandler requestHandler;
+
+  private final Editor editor;
+  private final ConversationService conversationService;
+  private final Conversation conversation;
+
+
+  public InlineCodeCompletionRenderer(Editor editor) {
+    this.editor = editor;
+    ((EditorEx) editor).getDocument().addDocumentListener(this);
+    editor.getContentComponent().addKeyListener(this);
+    editor.addEditorMouseListener(this);
+    editor.getCaretModel().addCaretListener(this);
+    editor.getSelectionModel().addSelectionListener(this);
+    initTypingTimer();
+    this.conversationService = ConversationService.getInstance();
+    var completionCode =
+        SettingsState.getInstance().getSelectedService().getCompletionCode() + ".inline";
+    this.conversation = conversationService.createConversation(completionCode);
+    conversationService.addConversation(this.conversation);
+  }
+
+  private void initTypingTimer() {
+    typingTimer = new Timer(ConfigurationState.getInstance().getInlineDelay(),
+        e -> ApplicationManager.getApplication().invokeLater(() -> {
+          if (editor != null && !editor.isDisposed()) {
+            Inlay<InlineCodeCompletionRenderer> inlay = editor.getUserData(INLAY_KEY);
+            if (inlay == null) {
+              Document document = editor.getDocument();
+              if (document.getText().isEmpty() || document.getText().isBlank()) {
+                return;
+              }
+              int offset = editor.getCaretModel().getOffset();
```

Comment:
What I mean by method/function level completion is that you split the method or function into two parts at the cursor's offset, and then send those two parts to the language model.

For example:
```java
var psiDocumentManager = PsiDocumentManager.getInstance(editor.getProject());
int cursorOffset = editor.getCaretModel().getOffset();
PsiFile psiFile = psiDocumentManager.getPsiFile(document);
if (psiFile != null) {
  PsiElement elementAtCaret = psiFile.findElementAt(cursorOffset);
  while (elementAtCaret != null) {
    if (elementAtCaret instanceof PsiMethod) {
      var method = (PsiMethod) elementAtCaret;
      var methodTextRangeInParent = method.getTextRangeInParent();
      String prefix = document.getText(new TextRange(methodTextRangeInParent.getStartOffset(), cursorOffset));
      String suffix = document.getText(new TextRange(cursorOffset, methodTextRangeInParent.getEndOffset()));

      // rest
      break;
    }
    elementAtCaret = elementAtCaret.getParent();
  }
}
```

Reply:
Yes I thought about that approach too, I just thought the code completion would be better if the LLM gets the context of the entire File.
I use the [CodeLlama-7B-Instruct-GGUF](https://huggingface.co/TheBloke/CodeLlama-7B-Instruct-GGUF) running behind a [llama-cpp-python OpenAI-API like Server](https://github.com/abetlen/llama-cpp-python?tab=readme-ov-file#openai-compatible-web-server) and with the current implementation I get the following completion in an old private Spring Boot project:
<details>
  <summary>MovieController.java</summary>

  ```java
  // imports ...

/**
 * REST Controller for Movies
 */
@RestController
@RequestMapping("/api/movies")
public class MovieController {
    private final MovieService movieService;
    private Logger log = LoggerFactory.getLogger(MovieController.class);

    @Autowired
    public MovieController(MovieService movieService) {
        this.movieService = movieService;
    }

    @GetMapping
    public Collection<Movie> getMovies(@RequestParam(value = "name", required = false) String search) {
        //-------------- Code from LLM:
         List<Movie> movies = null;
          if (null == search) {
              movies = new ArrayList<>();
          } else {
              movies = movieService.queryAllMovies(search);
          }
        return movies; 
        //-------------- 
    }

    @GetMapping("/page/{page}")
    public ResponseEntity<Collection<Movie>> getMoviesPaged(@RequestParam(value = "name", required = false) String search,
                                                            @PathVariable(value = "page") Integer page) {
        HttpHeaders responseHeaders = new HttpHeaders();
        Slice<Movie> slice;
        if (null == search) {
            slice = movieService.getAllMoviesPaged(page, 10);

        } else {
            slice = movieService.queryAllMoviesPaged(search, page, 10);
        }
        if (slice.hasNext()) {
            responseHeaders.set("hasMore", "true");
        } else {
            responseHeaders.set("hasMore", "false");
        }
        return ResponseEntity.ok().headers(responseHeaders).body(slice.getContent());
    }

    @GetMapping("/ids")
    public Map<Long, Movie> getMoviesList(@RequestParam(value = "ids") List<Long> movIds) {
        List<Movie> movies = movieService.queryMoviesByIds(movIds);
        Map<Long, Movie> moviesMap = movies.stream()
                .collect(Collectors.toMap(Movie::getMovId, Function.identity()));
        log.info("Map of Movies by ids: {}", Utils.printMap(moviesMap));
        return moviesMap;
    }

    @GetMapping("/{id}")
    ResponseEntity<Movie> getMovie(@PathVariable Long id) {
        Optional<Movie> movie = movieService.queryMovie(id);
        return movie.map(response -> ResponseEntity.ok().body(response))
                .orElse(new ResponseEntity<>(HttpStatus.NOT_FOUND));
    }

    @GetMapping("/trailer/{movId}")
    ResponseEntity<String> getMovieTrailer(@PathVariable Long movId) {
        Optional<String> trailer = movieService.getTrailer(movId);
        return trailer.map(response -> ResponseEntity.ok().body(response))
                .orElse(new ResponseEntity<>(HttpStatus.NOT_FOUND));
    }

    @PreAuthorize("hasRole('ADMIN')")
    @PostMapping
    ResponseEntity<?> createMovie(@RequestBody Movie movie) throws URISyntaxException {
        try {
            Movie result = movieService.saveMovieIfNotExists(movie);
            return ResponseEntity.created(new URI("/api/movie/" + result.getMovId()))
                    .body(result);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("{\"msg\":\"" + e.getMessage() + "\"}");
        }
    }

    @PreAuthorize("hasRole('ADMIN')")
    @PutMapping("/{id}")
    ResponseEntity<Movie> updateMovie(@Valid @RequestBody Movie movie) {
        Movie result = null;
        try {
            result = movieService.saveMovieIfNotExists(movie);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return ResponseEntity.ok().body(result);
    }

    @PreAuthorize("hasRole('ADMIN')")
    @DeleteMapping
    public ResponseEntity<?> deleteAllMovies() {
        movieService.deleteAll();
        return ResponseEntity.ok().build();
    }

}
  ```
</details>
Which does seem quite usable, and would probably be worse if the LLM gets only the enclosing method sent.
But I will try out your apporach and see what results I get.

Reply:
Really? I'm going to give it a try. OpenAI models seem to hallucinate A LOT, or perhaps I'm using some wrong settings.
Providing relevant context through the system prompt seems to make the results a bit better, but still the overall behaviour is not good.

Reply:
Was the completion result exactly the same as what you copy-pasted, which could be plugged directly into the function, or did it require some kind of human intervention? Because even with Code Llama 7B Q5, it doesn't look close to being usable.

![llama_1](https://github.com/carlrobertoh/CodeGPT/assets/15891408/fe0312d6-abfd-47e8-9e89-9c60f1d61a0c)
![llama_2](https://github.com/carlrobertoh/CodeGPT/assets/15891408/2802acdd-742e-4a7b-850c-14b61e0775eb)

Reply:
Sorry for the confusion, of course I used the non-instruct version [CodeLlama-7B-GGUF](https://huggingface.co/TheBloke/CodeLlama-7B-GGUF) (Q6).
The completion I posted is exactly what CodeLlama returned:
![image](https://github.com/carlrobertoh/CodeGPT/assets/39240633/dfb05156-5177-4ab1-a969-3096d0dd061d)
(`\n` does not seem to be rendered as an actual new line in the inlay)

But for your example it returns only an empty " <EOT>" for me. I tested it in Postman and for:
POST `/v1/chat/completions`
 ```json
 {
  "messages": [
    {
        "role": "user",
        "content": "<PRE> public static int binarySearch(int[] arr, int first, int last, int key) {\n <SUF> \n\n}\n} <MID>"
    }
  ],
  "temperature": 0.0,
  "maxTokens": 1000
}
 ```
It actually returns a correct implementation:
```json
{
  //...
    "choices": [
        {
            "index": 0,
            "message": {
                "content": "\n  if (first > last) return -1;\n  int mid = (first + last)/2;\n  if (arr[mid] == key) return mid;\n  else if (key < arr[mid]) return binarySearch(arr, first, mid-1, key);\n  else return binarySearch(arr, mid+1, last, key); <EOT>",
                "role": "assistant"
            },
            "finish_reason": "stop"
        }
    ],
}
```

But when the entire file is sent:
```json
```

Reply:
That was it! Sorry, my bad. I was using the instruct model that I had previously downloaded.
I'm also seeing the same behaviour with Q6 model when sending the entire file.

FYI: I'm going to start pushing some changes to your branch soon. Hopefully, there won't be many merge conflicts.

Reply:
@carlrobertoh I added a commit which adds the extraction of the correct completion prompt if the cursor is enclosed by a Method or Class. It now only sends the entire file content if there is no Method or Class found around the cursor.

https://github.com/carlrobertoh/CodeGPT/blob/26408382f16287647e51b9226b3f2d85ebe3b7cd/src/main/java/ee/carlrobert/codegpt/codecompletions/CodeCompletionService.java#L129-L135

Reply:
@PhilKes I don't think this will work properly because the code completion will be triggered regardless of the cursor's position within the class. This is because the logic iterates through the PSI hierarchy until it finds a PsiClass instance.

Reply:
@carlrobertoh I am not sure how you mean that. If the cursor is inside a method, the `tryFindEnclosingPsiElement` will find the enclosing `PsiMethod` first and return it, if the cursor is placed outside of any method but inside a class it will find a `PsiClass`. I tested this behaviour in Java.

Reply:
@PhilKes Yes, the behavior itself works, but the problem lies in determining when to trigger the completion. I don't think we can control this by comparing the correct PsiElement instances where the cursor is located, or at least this won't be enough. 

In general, we want to prevent code completion from triggering when the cursor is in the middle of a keyword or identifier, or in areas where the language semantics do not allow for any meaningful completions.

I'm not yet sure how we can achieve that or how to determine the correct context, but I believe we can improve it gradually.

Reply:
@carlrobertoh  I see what you mean, it will be quite tricky to find some conditions that fit all programming langauges.
A first approach I think would be:
```java
  PsiElement elementAtCaret = psiFile.findElementAt(caretOffset);
  if (!(elementAtCaret instanceof PsiWhiteSpace)) {
    return; // skip completion
  }
```
This would at least ensure that the completion is only triggered if the cursor is not inside some existing identifier/keyword:
```java
public int binarySearch(int[] arr, int first, int last, int key) {
       ret{CURSOR_HERE}urn  // would not trigger completion
    }

public int binary{<CURSOR_HERE}Search(int[] arr, int first, int last, int key) {  // would not trigger completion
       return
    }

public int binarySearch(int[] arr, int first, int last, int key) { 
       return{CURSOR_HERE} // would trigger the completion
    }

public int binarySearch(int[] arr, int{CURSOR_HERE} first, int last, int key) {   // would trigger the completion (not wanted)
       return
    }
```
We could try to evalute `elementAtCaret.getNextSibling()` and `elementAtCaret.getPrevSibling()` too, for the last example we could say if the previous Element is a `PsiTypeElement` and the next Element is a `PsiIdentifier` than the cursor is inside of a variable declaration, but that is only correct for programming languages where you declare variables like that.

Reply:
Yep, this will do for now. In the future, we might need to look into [LSP](https://microsoft.github.io/language-server-protocol/) or [tree-sitter](https://tree-sitter.github.io/tree-sitter/), which should provide greater control over the code context.